File: tc-asn.py
#!/usr/bin/python

import dns.resolver
import dns.reversename
import fileinput
import re
import sys

def get_ptr_name(addr):
    """
    Return a Cymru BGP lookup PTR name given an IPv4 or IPv6 address
    """

    # use library to get reverse name
    name = dns.reversename.from_address(addr).to_text(omit_final_dot=True)
    # remove in-addr.arpa or ip6.arpa suffix
    name = name.rsplit('.', 2)[0]

    # append v4 or v6 suffix depending on ip version
    if ':' in addr:
        name += '.origin6.asn.cymru.com.'
    else:
        name += '.origin.asn.cymru.com.'

    return name

def get_asn(addr):
    """
    Return a list of origin ASNs given an IPv4 or IPv6 address
    """

    qname = get_ptr_name(addr)

    try:
        answer = dns.resolver.query(qname, rdtype='TXT', rdclass='IN')
    except:
        # TODO: send this to stderr, log, or use traceback.format_exc()
        return ['NA']

    asns = []
    for rdata in answer:
        # https://www.cymru.com/jtk/blog/2010/08/10/#parsing-tcbgp-mapping
        #   "49152 [...] | 192.0.2.0/24 | AA | registry | 1970-01-01"
        #   [...]
        # parse each TXT line, extract one or more ASNs per line
        rdatatxt = rdata.strings
        for line in rdatatxt:
            asnstr = [x.strip() for x in line.split('|')]
            asnlist = asnstr[0].split()
            asns.extend(asnlist)

    # remove duplicate ASNs, then return the list
    asns = list(set(asns))
    return asns

# expect IP addrs, one per line
for line in fileinput.input():
    line = line.rstrip()

    asns = get_asn(line)
    for asn in asns:
        sys.stdout.write('%-10s | %s\n' % (asn, line))

File: tc-asn.pl
#!/usr/bin/perl -T
use strict;
use warnings;
$|=1;

# $Id: sample-tcbgp-mapping.pl,v 1.1 2010/08/10 16:34:28 jtk Exp $
# sample code using the Team Cymru IP address to BGP DNS-based mapping service
# Also see: <http://www.cymru.com/jtk/blog/2010/08/10/#parsing-tcbgp-mapping>
# WARNING: not a complete, robust service interface, see blog for details

use Net::DNS;
use Net::IP;

while ( defined(my $line=<>) ) {
    chomp $line;
    printf "%-11s  |  %s\n", get_asn($line) || 'NA', $line;
}

sub get_asn {
    my $address = shift || return;
    my $res     = Net::DNS::Resolver->new;
    my $qname   = get_ptr_name($address);
    my $query   = $res->send( $qname, 'TXT', 'IN' );
    my $asn;

    return if !$query;
    return if $query->header->ancount < 1;

ANSWER:
    for my $answer ( $query->answer ) {
        next ANSWER if $answer->type ne 'TXT';
        ($asn) = $answer->rdatastr =~ m{ \A ["] (\d+) }xms;
        $asn ? last ANSWER : next ANSWER;
     }

    return $asn;
}

sub get_ptr_name {
    my $addr = shift || return;

    if ( $addr =~ /:/ ) {
        $addr  = substr new Net::IP ($addr)->reverse_ip, 0, -10;
        $addr .= '.origin6.asn.cymru.com';
    }
    else {
        $addr  = join( '.', reverse split( /\./, $addr ) );
        $addr .=  '.origin.asn.cymru.com';
    }
    return $addr;
}
